File: src/main/java/io/github/przbetkier/tuscan/client/lambda/LambdaDemoParserClient.java
package io.github.przbetkier.tuscan.client.lambda;

import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.http.HttpMethod;
import org.springframework.http.MediaType;
import org.springframework.stereotype.Component;
import org.springframework.web.reactive.function.BodyInserters;
import org.springframework.web.reactive.function.client.WebClient;
import reactor.core.publisher.Mono;

@Component
public class LambdaDemoParserClient {

    private final WebClient lambdaClient;

    public LambdaDemoParserClient(@Qualifier("awsLambdaClient") WebClient lambdaClient) {
        this.lambdaClient = lambdaClient;
    }

    public Mono<LambdaResponse> invokeLambda(LambdaRequest lambdaRequest) {
        return lambdaClient.method(HttpMethod.POST)
                .uri("/default/demo-parser")
                .contentType(MediaType.APPLICATION_JSON)
                .body(BodyInserters.fromValue(lambdaRequest))
                .retrieve()
                .bodyToMono(LambdaResponse.class);
    }
}

File: src/main/java/pro/tuscan/client/lambda/LambdaDemoParserClient.kt
package pro.tuscan.client.lambda

import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.http.HttpMethod.POST
import org.springframework.http.MediaType.APPLICATION_JSON
import org.springframework.stereotype.Component
import org.springframework.web.reactive.function.BodyInserters.fromValue
import org.springframework.web.reactive.function.client.WebClient
import pro.tuscan.adapter.api.LambdaRequest
import pro.tuscan.adapter.api.LambdaResponse
import reactor.core.publisher.Mono

@Component
class LambdaDemoParserClient(@Qualifier("awsLambdaClient") private val lambdaClient: WebClient) {

    fun invokeLambda(lambdaRequest: LambdaRequest): Mono<LambdaResponse> =
            lambdaClient.method(POST)
                    .uri("/default/demo-parser")
                    .contentType(APPLICATION_JSON)
                    .body(fromValue(lambdaRequest))
                    .retrieve()
                    .bodyToMono(LambdaResponse::class.java)
}

